flink-end-to-end-perf-tests/flink-basic-operations/src/main/java/org/apache/flink/basic/utils/FromElements.java

```diff
@@ -0,0 +1,229 @@
+		}
+		Collection<OUT> dataCollection = Arrays.asList(data);
+		FromElementsRichFunction.checkCollection(dataCollection, typeInfo.getTypeClass());
+		try {
+			SourceFunction<OUT> function =
+				new FromElementsRichFunction<>(typeInfo.createSerializer(env.getConfig()),
+					maxCount, sleepNum, dataCollection);
+			return env.addSource(function, "WordsData", typeInfo);
+		} catch (IOException e) {
+			throw new RuntimeException(e.getMessage(), e);
+		}
+	}
+
+	/**
+	 * A stream source function that returns a sequence of elements.
+	 *
+	 * <p>
+	 * Upon construction, this source function serializes the elements using Flink's type information. That way, any
+	 * object transport using Java serialization will not be affected by the serializability of the elements.
+	 * </p>
+	 * <p>
+	 * <b>NOTE:</b> This source has a parallelism of 1.
+	 * </p>
+	 *
+	 * @param <T> The type of elements returned by this function.
+	 */
+	public static class FromElementsRichFunction<T> extends RichSourceFunction<T>
+		implements ParallelSourceFunction<T>, CheckpointedFunction {
+		private static final long serialVersionUID = 1L;
+		private final TypeSerializer<T> serializer;
+		private final byte[] elementsSerialized;
+		private volatile long numElementsEmitted;
+		private volatile long numElementsToSkip;
+		private long recordCount = 0L;
+		private long maxCount = 0L;
+		private int sleepNum = 0;
+		private transient ListState<Long> operateState;
+		private transient Meter sourceTpsMetrics;
+		private volatile boolean isRunning = true;
+
+		public FromElementsRichFunction(TypeSerializer<T> serializer, Long maxCount, int sleepNum, Iterable<T> elements) throws IOException {
+			ByteArrayOutputStream baos = new ByteArrayOutputStream();
+			DataOutputViewStreamWrapper wrapper = new DataOutputViewStreamWrapper(baos);
+			try {
+				for (T element : elements) {
+					serializer.serialize(element, wrapper);
+				}
+			} catch (Exception e) {
+				throw new IOException("Serializing the source elements failed: " + e.getMessage(), e);
+			}
+			this.serializer = serializer;
+			this.elementsSerialized = baos.toByteArray();
+			this.maxCount = maxCount;
+			this.sleepNum = sleepNum;
+		}
+
+		@Override
+		public void open(Configuration parameters) throws Exception {
+			RuntimeContext runtimeContext = getRuntimeContext();
+			MetricGroup metricGroup = runtimeContext.getMetricGroup();
+			final String sourceTpsName = "tps";
+			Counter sourceCounter = metricGroup.counter(sourceTpsName + "_counter", new SimpleCounter());
+			sourceTpsMetrics = metricGroup.meter(sourceTpsName, new MeterView(sourceCounter, 60));
+		}
+
+		@Override
+		public void close() {}
+
+		@Override
+		public void initializeState(FunctionInitializationContext context) throws Exception {
+			Preconditions.checkState(this.operateState == null,
+				"The " + getClass().getSimpleName() + " has already been initialized.");
+			this.operateState = context.getOperatorStateStore().getOperatorState(
```

Comment:
`org.apache.flink.api.common.state.OperatorStateStore#getOperatorState` is deprecated